enhancements/network/on-prem-mutable-vips.md

```diff
@@ -0,0 +1,213 @@
+
+### Drawbacks
+
+Because we do not implement Update operation, the main drawback is a scenario when user adds an incorect address to be configured. We will only allow deleting a second entry, therefore user will need to first delete and then add again a correct pair of VIPs. Because fixing such a typo will require two reboots (one for deleting and one for adding),
+enablement of this feature should be performed with some level of care by the end user.
+
+## Design Details
+
+This feature will require changes in a few different components:
+
+* Installer - New fields will need to be populated during the installation. This is a safe part as installer is already populating `status` and now will populate `status` and `spec`.
+
+* API - New fields will need to be added to the platform spec section of the infrastructure object.
+
+* Cluster Network Operator - New object will need to be watched by one of its controllers. The core logic of validating the requested change will be implemented there.
+
+* Machine Config Operator - Definition of the runtimecfg Pod will need to be rendered again when the VIP configuration changes.
+
+* baremetal-runtimecfg - The code rendering the keepalived template belongs to this repo. As the enhancement does not touch the rendering part no big changes are expected in the runtimecfg. This component is used to re-render configuration in case something changes in the cluster (e.g. new nodes are added) so the fact we modify configuration is not a new feature from its perspective.
+Small changes may be needed to acommodate to the new user scenario to provide good user experience.
+
+### openshift/api
+
+The new structure of platformSpec would look like this:
+
+```yaml
+platformSpec:
+  properties:
+    baremetal:
+      description: BareMetal contains settings specific to the BareMetal platform.
+      type: object
+      properties:
+        apiServerInternalIPs:
+          description: apiServerInternalIPs are the IP addresses to...
+          type: array
+        ingressIPs:
+          description: ingressIPs are the external IPs which...
+          type: array
+      ...
+```
+
+This format (i.e. usage of arrays) is already used by platformStatus so this does not introduce a new type nor schema.
+
+### Cluster Network Operator
+
+There already exists `infrastructureconfig_controller` inside CNO that watches for changes in `configv1.Infrastructure` for on-prem platforms. Because of this we do not need to create a new one nor change any fundaments, we only need to extend its logic.
+
+Currently the controller already reconciles on the `CreateEvent` and `UpdateEvent` so we will implement a set of functions that compare and validate `platformSpec` with `platformStatus` and do what's needed.
+
+### Machine Config Operator
```

Comment:
Will MCO be the component responsible for setting the updated `PlatformStatus`?

Reply:
CNO will be responsible for syncing the spec and status.

Reply:
Tell me if I get the question wrong, but from https://github.com/openshift/enhancements/pull/1463#discussion_r1316448073

* Installer sets the initial values at install-time
* CNO sets the values if changed as cluster lifecycle operation

MCO does not own any CR here, it is only responsible for applying the change on the filesystem (i.e. templating the config file)